File: Problems & Solutions/1519. Number of Nodes in the Sub-Tree With the Same Label/12 January 2022/Approach 2 - Breadth First Search/Number of Nodes in the Sub-Tree With the Same Label.cpp
class Solution {
public:
    vector<int> countSubTrees(int n, vector<vector<int>>& edges, string labels) {
        unordered_map<int, unordered_set<int>> adj;
        for (auto& edge : edges) {
            adj[edge[0]].insert(edge[1]);
            adj[edge[1]].insert(edge[0]);
        }

        vector<vector<int>> counts(n, vector<int>(26));
        queue<int> q;

        for (int node = 0; node < n; ++node) {
            counts[node][labels[node] - 'a'] = 1;
            if (node != 0 && adj[node].size() == 1) {
                q.push(node);
            }
        }

        while (q.size()) {
            int curr = q.front();
            q.pop();

            int parent = *adj[curr].begin();
            adj[parent].erase(curr);

            for (int i = 0; i < 26; ++i) {
                counts[parent][i] += counts[curr][i];
            }

            if (parent != 0 && adj[parent].size() == 1) {
                q.push(parent);
            }
        }

        vector<int> ans(n);
        for (int node = 0; node < n; ++node) {
            ans[node] = counts[node][labels[node] - 'a'];
        }

        return ans;
    }
};

File: Problems & Solutions/1519. Number of Nodes in the Sub-Tree With the Same Label/12 January 2022/Approach 2 - Breadth First Search/Number of Nodes in the Sub-Tree With the Same Label.java
class Solution {
    public int[] countSubTrees(int n, int[][] edges, String labels) {
        Map<Integer, Set<Integer>> adj = new HashMap<>();
        for (int[] edge : edges) {
            int a = edge[0], b = edge[1];
            adj.computeIfAbsent(a, value -> new HashSet<Integer>()).add(b);
            adj.computeIfAbsent(b, value -> new HashSet<Integer>()).add(a);
        }
        int[][] counts = new int[n][26];
        Queue<Integer> q = new LinkedList<>();

        for (int i = 0; i < n; i++) {
            counts[i][labels.charAt(i) - 'a'] = 1;
            if (i != 0 && adj.get(i).size() == 1) {
                q.offer(i);
            }
        }

        while (!q.isEmpty()) {
            int node = q.poll();

            int parent = adj.get(node).stream().findFirst().get();
            adj.get(parent).remove(node);

            for (int i = 0; i < 26; i++) {
                counts[parent][i] += counts[node][i];
            }

            if (parent != 0 && adj.get(parent).size() == 1) {
                q.offer(parent);
            }
        }

        int[] ans = new int[n];
        for (int i = 0; i < n; i++) {
            ans[i] = counts[i][labels.charAt(i) - 'a'];
        }
        return ans;
    }
}